### PR TITLE
Firefox: Pass page wait request

### PR DESCRIPTION
**Pull request by Meir017**
part of #113

based on #295

### CODE REVIEW

**Comment by Meir017** [comment]
another issue with codeFactor - `Page.WaitForEvent` is a switch-case

**Comment by kblok** [comment]
Same title as #295?

**Comment by Meir017** [comment]
@kblok not exactly, this doesn't cover waiting for request/response events

**Comment by kblok** [comment]
conflicted

File: src/PlaywrightSharp.Tests/Page/Events/PageEventsConsoleTests.cs [hunk 1]
```diff
@@ -1,6 +1,7 @@
 using System.Collections.Generic;
 using System.Linq;
 using System.Threading.Tasks;
+using PlaywrightSharp.Helpers;
```
Comment (kblok) [inline src/PlaywrightSharp.Tests/Page/Events/PageEventsConsoleTests.cs:4]: Do we need this?

File: src/PlaywrightSharp/Frame.cs [hunk 2]
```diff
@@ -1,543 +1,562 @@
-using System;
```
Comment (kblok) [inline src/PlaywrightSharp/Frame.cs:1]: Ops here again

File: src/PlaywrightSharp/FrameManager.cs [hunk 3]
```diff
@@ -181,6 +181,41 @@ internal void ClearFrameLifecycle(Frame frame)
             }
         }
 
+        internal void RequestStarted(Request request)
+        {
+            InflightRequestStarted(request);
+            var frame = request.Frame;
+            if (request.DocumentId != null && frame != null && request.RedirectChain.Length == 0)
```
Comment (kblok) [inline src/PlaywrightSharp/FrameManager.cs:188]: `string.IsNullOrEmpty` for `DocumentId`?

File: src/PlaywrightSharp/Page.cs [hunk 4]
```diff
@@ -356,8 +356,61 @@ public Task<IPage> GetOpenerAsync()
...
+                        var genericEvent = (T)(object)e;
+                        if (options?.Predicate != null && options.Predicate(genericEvent))
+                        {
+                            requestTsc.SetResult(genericEvent);
+                            Request -= RequestEventHandler;
+                        }
+                    }
+
+                    Request += RequestEventHandler;
+                    return await requestTsc.Task.WithTimeout(timeout).ConfigureAwait(false);
```
Comment (kblok) [inline src/PlaywrightSharp/Page.cs:387]: Can we make this a method?

File: src/PlaywrightSharp/Page.cs [hunk 5]
```diff
@@ -356,8 +356,61 @@ public Task<IPage> GetOpenerAsync()
...
+                        var genericEvent = (T)(object)e;
+                        if (options?.Predicate != null && options.Predicate(genericEvent))
+                        {
+                            responseTsc.SetResult(genericEvent);
+                            Response -= ResponseEventHandler;
+                        }
+                    }
+
+                    Response += ResponseEventHandler;
+                    return await responseTsc.Task.WithTimeout(timeout).ConfigureAwait(false);
```
Comment (kblok) [inline src/PlaywrightSharp/Page.cs:403]: Same

File: src/PlaywrightSharp/Frame.cs (hunk 2 above)
Comment (Meir017) [reply src/PlaywrightSharp/Frame.cs:1]: not sure how to fix this

File: src/PlaywrightSharp/Page.cs (hunk 4 above)
Comment (Meir017) [reply src/PlaywrightSharp/Page.cs]: done

File: src/PlaywrightSharp/Page.cs [hunk 6]
```diff
@@ -357,6 +361,47 @@ public Task<IPage> GetOpenerAsync()
 
         /// <inheritdoc cref="IPage.WaitForEvent{T}(PageEvent, WaitForEventOptions{T})"/>
         public Task<T> WaitForEvent<T>(PageEvent e, WaitForEventOptions<T> options = null)
+        {
+            ValidateArgumentsTypes();
+            var info = _pageEventsMap[e];
```
Comment (Meir017) [inline src/PlaywrightSharp/Page.cs:366]: @kblok check this out

File: src/PlaywrightSharp/Page.cs [hunk 7]
```diff
@@ -15,6 +16,9 @@ public sealed class Page : IPage, IDisposable
     {
         private readonly TaskCompletionSource<bool> _closeTsc = new TaskCompletionSource<bool>();
         private readonly Dictionary<string, Delegate> _pageBindings = new Dictionary<string, Delegate>();
+        private readonly Dictionary<PageEvent, EventInfo> _pageEventsMap = ((PageEvent[])Enum.GetValues(typeof(PageEvent)))
+            .ToDictionary(x => x, x => typeof(Page).GetEvent(x.ToString()));
```
Comment (kblok) [inline src/PlaywrightSharp/Page.cs:20]: This could be static

File: src/PlaywrightSharp/Page.cs [hunk 8]
```diff
@@ -357,6 +361,47 @@ public Task<IPage> GetOpenerAsync()
 
         /// <inheritdoc cref="IPage.WaitForEvent{T}(PageEvent, WaitForEventOptions{T})"/>
         public Task<T> WaitForEvent<T>(PageEvent e, WaitForEventOptions<T> options = null)
+        {
+            ValidateArgumentsTypes();
+            var info = _pageEventsMap[e];
+            var eventTsc = new TaskCompletionSource<T>();
+            void PageEventHandler(object sender, T e)
+            {
+                if (options?.Predicate == null || options.Predicate(e))
```
Comment (kblok) [inline src/PlaywrightSharp/Page.cs:370]:
```suggestion
                if (options?.Predicate?.Invoke(e) == true)

```

File: src/PlaywrightSharp/Page.cs [hunk 9]
```diff
@@ -637,6 +675,38 @@ private async Task ExposeFunctionAsync(string name, Delegate puppeteerFunction)
             await Delegate.ExposeBindingAsync(name, GetEvaluationString(addPageBinding, name)).ConfigureAwait(false);
         }
 
+        private Task<RequestEventArgs> WaitForRequestInternalAsync(Func<RequestEventArgs, bool> predicate, int? timeout = null)
+        {
+            var requestTsc = new TaskCompletionSource<RequestEventArgs>();
+            void RequestEventHandler(object sender, RequestEventArgs e)
+            {
+                if (predicate == null || predicate(e))
```
Comment (kblok) [inline src/PlaywrightSharp/Page.cs:683]:
```suggestion
                if (predicate?.Invoke(e) == true)

```

File: src/PlaywrightSharp/Page.cs [hunk 10]
```diff
@@ -637,6 +675,38 @@ private async Task ExposeFunctionAsync(string name, Delegate puppeteerFunction)
...
+            Request += RequestEventHandler;
+            return requestTsc.Task.WithTimeout(timeout ?? DefaultTimeout);
+        }
+
+        private Task<ResponseEventArgs> WaitForResponseInternalAsync(Func<ResponseEventArgs, bool> predicate, int? timeout)
+        {
+            var tsc = new TaskCompletionSource<ResponseEventArgs>();
+            void ResponseEventHandler(object sender, ResponseEventArgs e)
+            {
+                if (predicate == null || predicate(e))
```
Comment (kblok) [inline src/PlaywrightSharp/Page.cs:699]:
```suggestion
                if (predicate?.Invoke(e) == true)

```

File: src/PlaywrightSharp/Page.cs (hunk 8 above)
Comment (Meir017) [reply src/PlaywrightSharp/Page.cs:370]: not sure this is equivalent to what we want. if `options.Predicate` is not given then we want to enter the if statement

Comment (kblok) [reply src/PlaywrightSharp/Page.cs:370]: Good point :+1:

**Pull request merged**